### PR TITLE
[FIXED JENKINS-45448] Add a display name to GitHub Servers configuration

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-45448](https://issues.jenkins-ci.org/browse/JENKINS-45448)

@reviewbybees

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/172)

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java [hunk 1]
```diff
@@ -127,6 +162,19 @@ public void setManageHooks(boolean manageHooks) {
     public void setCustomApiUrl(boolean customApiUrl) {
     }
 
+    /**
+     * Gets the optional display name of this server.
+     * @return the optional display name of this server, may be empty or {@code null} but best effort is made to ensure
+     * that it has some meaningful text.
+     */
+    public String getName() {
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: `getDisplayName()` is standard pattern for jenkins where something should be shown in UI

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java [hunk 2]
```diff
@@ -59,22 +62,45 @@
...
+     *
+     * @since 1.27.0
+     */
+    private static final String[] COMMON_PREFIX_HOSTNAMES = {
+        "git.",
+        "github.",
+        "vcs.",
+        "scm.",
+        "source."
+    };
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: why it needed?

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: Because if somebody provides an url like `https://github.example.com` we really would rather give that the name `example` rather than `github example` because the `github` is redundant information... on the other hand if it was `https://github1.example.com` then we want `github1 example` as the suggested display name

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java (hunk 1 above)
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: I think we can use both - `getDisplayName` and `getName` as `getName` looks like simple getter, but inside contains some logic (which should be avoided for getters).
With displayName we can emphasize that it gets name or some calculated default value

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java [hunk 3]
```diff
@@ -59,22 +62,45 @@
...
      * Because of {@link GitHub} hide this const from external use we need to store it here
      */
     public static final String GITHUB_URL = "https://api.github.com";
 
     /**
+     * The name to display for the public GitHub service.
+     *
+     * @since 1.27.0
+     */
+    private static final String GITHUB_NAME = "GitHub";
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: Maybe "Public GitHub"?

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: So to clarify I have added the `getDisplayName()` method that produces the formatted display name

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java [hunk 4]
```diff
@@ -127,6 +161,34 @@ public void setManageHooks(boolean manageHooks) {
...
 
+    /**
+     * Gets the optional display name of this server.
+     *
+     * @return the optional display name of this server, may be empty or {@code null} but best effort is made to ensure
+     * that it has some meaningful text.
+     * @since 1.28.0
+     */
+    public String getName() {
+        if (StringUtils.isBlank(name)) {
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: with new display name method it can be simple getter

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java (hunk 3 above)
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: exactly what i wanted comment

**Comment by KostyaSha** [comment]
some push hide our comments

File: src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java [hunk 5]
```diff
@@ -184,8 +180,11 @@ public String getName() {
      */
     public String getDisplayName() {
         String n = getName();
-        String a = StringUtils.isBlank(apiUrl) || GITHUB_URL.equals(apiUrl)
-                ? "https://github.com" : StringUtils.removeEnd(apiUrl, "/api/v3");
+        boolean gitHubOrg = StringUtils.isBlank(apiUrl) || GITHUB_URL.equals(apiUrl);
+        if (StringUtils.isBlank(n)) {
+            n = gitHubOrg ? GITHUB_NAME : SCMName.fromUrl(apiUrl, COMMON_PREFIX_HOSTNAMES);
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/config/GitHubServerConfig.java]: please avoid one-letter vars next time. In java its very hard to follow code if it has obfuscated variables

**Comment by lanwen** [comment]
what about `PUBLIC_GITHUB`?

**Comment by lanwen** [comment]
hmm it's strange, but where CI build? @oleg-nenashev

**Comment by KostyaSha** [comment]
@lanwen probably they disabled standard builder and enforcing everybody to use their cloudbees pipeline

**Comment by lanwen** [comment]
meh... is there any info about this change?

**Comment by KostyaSha** [comment]
We can add travis build ;)

**Comment by lanwen** [comment]
Travis hangs on tests :( i've tried
